File: node/cogs/daily_announcement.js
const {
  SlashCommandBuilder,
  AttachmentBuilder,
  EmbedBuilder,
  ActionRowBuilder,
  StringSelectMenuBuilder,
} = require('discord.js');
const fs = require('fs');
const path = require('path');
const { mkdir } = require('fs/promises');
const translate = require('@vitalets/google-translate-api');

const IMAGE_DIR = path.join('images', 'daily');
const DAILY_SELECT_ID = 'daily:select';

const createCommand = new SlashCommandBuilder()
  .setName('aviso_diario')
  .setDescription('Cria um aviso diário (máx 4 ativos)')
  .addStringOption((option) =>
    option
      .setName('texto_pt')
      .setDescription('Texto em português')
      .setRequired(true),
  )
  .addAttachmentOption((option) =>
    option
      .setName('imagem_pt')
      .setDescription('Imagem em português')
      .setRequired(true),
  )
  .addAttachmentOption((option) =>
    option
      .setName('imagem_en')
      .setDescription('Imagem em inglês')
      .setRequired(true),
  );

const disableCommand = new SlashCommandBuilder()
  .setName('aviso_diario_desativar')
  .setDescription('Desativa um aviso diário específico');

async function downloadAttachment(attachment, targetPath) {
  const response = await fetch(attachment.url);
  if (!response.ok) {
    throw new Error(`Falha ao baixar ${attachment.url}`);
  }
  const buffer = Buffer.from(await response.arrayBuffer());
  await fs.promises.writeFile(targetPath, buffer);
}

async function handleCreate({ interaction, db }) {
  await interaction.deferReply({ ephemeral: true });

  const active = await db.getActiveDailyAnnouncements();
  if (active.length >= 4) {
    await interaction.editReply('❌ Limite de 4 avisos ativos atingido.');
    return;
  }

  const textoPt = interaction.options.getString('texto_pt', true);
  const imagemPt = interaction.options.getAttachment('imagem_pt', true);
  const imagemEn = interaction.options.getAttachment('imagem_en', true);

  if (!imagemPt.contentType?.startsWith('image') || !imagemEn.contentType?.startsWith('image')) {
    await interaction.editReply('❌ Os anexos precisam ser imagens.');
    return;
  }

  const translation = await translate(textoPt, { to: 'en' });
  const textoEn = translation.text;

  await mkdir(IMAGE_DIR, { recursive: true });
  const ts = Math.floor(Date.now() / 1000);
  const imgPtPath = path.join(IMAGE_DIR, `${ts}_pt.png`);
  const imgEnPath = path.join(IMAGE_DIR, `${ts}_en.png`);

  await downloadAttachment(imagemPt, imgPtPath);
  await downloadAttachment(imagemEn, imgEnPath);

  await db.addDailyAnnouncement(textoPt, textoEn, imgPtPath, imgEnPath);

  await interaction.editReply('✅ Aviso diário criado com sucesso.');
}

function buildDisableSelect(announcements) {
  return new ActionRowBuilder().addComponents(
    new StringSelectMenuBuilder()
      .setCustomId(DAILY_SELECT_ID)
      .setPlaceholder('Selecione o aviso para desativar')
      .addOptions(
        announcements.map((ann) => ({
          label: ann.text_pt.slice(0, 80),
          value: String(ann.id),
        })),
      ),
  );
}

async function handleDisable({ interaction, db }) {
  const announcements = await db.getActiveDailyAnnouncements();
  if (announcements.length === 0) {
    await interaction.reply({ content: '❌ Nenhum aviso ativo.', ephemeral: true });
    return;
  }

  await interaction.reply({
    content: 'Selecione o aviso que deseja desativar:',
    components: [buildDisableSelect(announcements)],
    ephemeral: true,
  });
}

function startDailyLoop({ client, db, forceChannelId }) {
  const interval = setInterval(async () => {
    const now = new Date();
    if (now.getMinutes() !== 0) {
      return;
    }

    if (now.getHours() < 12 || now.getHours() > 15) {
      return;
    }

    const index = now.getHours() - 12;
    const announcements = await db.getActiveDailyAnnouncements();
    if (index >= announcements.length) {
      return;
    }

    const channel = await client.channels.fetch(forceChannelId).catch(() => null);
    if (!channel) {
      return;
    }

    const ann = announcements[index];
    const filePt = new AttachmentBuilder(ann.image_pt_path, { name: 'pt.png' });
    const fileEn = new AttachmentBuilder(ann.image_en_path, { name: 'en.png' });

    const embedPt = new EmbedBuilder()
      .setTitle('📢 Aviso')
      .setDescription(ann.text_pt)
      .setColor(0x2ecc71)
      .setImage('attachment://pt.png');

    const embedEn = new EmbedBuilder()
      .setTitle('📢 Notice')
      .setDescription(ann.text_en)
      .setColor(0x3498db)
      .setImage('attachment://en.png');

    await channel.send({ embeds: [embedPt, embedEn], files: [filePt, fileEn] });
  }, 60 * 1000);

  return () => clearInterval(interval);
}

function registerDailyAnnouncement({ client, db, forceChannelId }) {
  startDailyLoop({ client, db, forceChannelId });

  client.on('interactionCreate', async (interaction) => {
    if (interaction.isStringSelectMenu()) {
      if (interaction.customId !== DAILY_SELECT_ID) {
        return;
      }

      const annId = Number(interaction.values[0]);
      await db.deactivateDailyAnnouncement(annId);
      await interaction.reply({
        content: '✅ Aviso desativado com sucesso.',
        ephemeral: true,
      });
      return;
    }

    if (!interaction.isChatInputCommand()) {
      return;
    }

    if (interaction.commandName === 'aviso_diario') {
      await handleCreate({ interaction, db });
      return;
    }

    if (interaction.commandName === 'aviso_diario_desativar') {
      await handleDisable({ interaction, db });
    }
  });
}

module.exports = {
  commands: [createCommand, disableCommand],
  registerDailyAnnouncement,
};

File: cogs/daily_announcement.py
import datetime
import os
import tempfile
import time

import discord
from deep_translator import GoogleTranslator
from discord import app_commands
from discord.ext import commands, tasks

import db
from utils.image_storage import is_remote_storage_enabled, is_remote_url, upload_image

FORCE_CHANNEL_ID = 1431340166877806644
IMAGE_DIR = "images/daily"


class AnnouncementSelect(discord.ui.Select):
    def __init__(self, announcements):
        options = [
            discord.SelectOption(label=ann[1][:80], value=str(ann[0]))
            for ann in announcements
        ]

        super().__init__(
            placeholder="Selecione o aviso para desativar", options=options
        )

    async def callback(self, interaction: discord.Interaction):
        ann_id = int(self.values[0])
        db.deactivate_daily_announcement(ann_id)
        await interaction.response.send_message(
            "✅ Aviso desativado com sucesso.", ephemeral=True
        )


class AnnouncementView(discord.ui.View):
    def __init__(self, announcements):
        super().__init__(timeout=60)
        self.add_item(AnnouncementSelect(announcements))


class DailyAnnouncement(commands.Cog):
    def __init__(self, bot):
        self.bot = bot
        os.makedirs(IMAGE_DIR, exist_ok=True)
        self.hourly_loop.start()

    # =========================
    # CREATE ANNOUNCEMENT
    # =========================
    @app_commands.command(
        name="aviso_diario", description="Cria um aviso diário (máx 4 ativos)"
    )
    async def aviso_diario(
        self,
        interaction: discord.Interaction,
        texto_pt: str,
        imagem_pt: discord.Attachment,
        imagem_en: discord.Attachment,
    ):
        await interaction.response.defer(ephemeral=True)

        active = db.get_active_daily_announcements()
        if len(active) >= 4:
            await interaction.followup.send(
                "❌ Limite de 4 avisos ativos atingido.", ephemeral=True
            )
            return

        if not imagem_pt.content_type.startswith(
            "image"
        ) or not imagem_en.content_type.startswith("image"):
            await interaction.followup.send(
                "❌ Os anexos precisam ser imagens.", ephemeral=True
            )
            return

        texto_en = GoogleTranslator(source="pt", target="en").translate(texto_pt)

        ts = int(time.time())

        if is_remote_storage_enabled():
            with tempfile.TemporaryDirectory(prefix="daily_announcement_") as temp_dir:
                local_pt = os.path.join(temp_dir, "pt.png")
                local_en = os.path.join(temp_dir, "en.png")
                await imagem_pt.save(local_pt)
                await imagem_en.save(local_en)

                img_pt_path = upload_image(local_pt, f"daily_{ts}_pt.png")
                img_en_path = upload_image(local_en, f"daily_{ts}_en.png")
        else:
            img_pt_path = f"{IMAGE_DIR}/{ts}_pt.png"
            img_en_path = f"{IMAGE_DIR}/{ts}_en.png"
            await imagem_pt.save(img_pt_path)
            await imagem_en.save(img_en_path)

        db.add_daily_announcement(texto_pt, texto_en, img_pt_path, img_en_path)

        await interaction.followup.send(
            "✅ Aviso diário criado com sucesso.", ephemeral=True
        )

    # =========================
    # DISABLE ANNOUNCEMENT
    # =========================
    @app_commands.command(
        name="aviso_diario_desativar", description="Desativa um aviso diário específico"
    )
    async def aviso_diario_desativar(self, interaction: discord.Interaction):
        anns = db.get_active_daily_announcements()

        if not anns:
            await interaction.response.send_message(
                "❌ Nenhum aviso ativo.", ephemeral=True
            )
            return

        view = AnnouncementView(anns)
        await interaction.response.send_message(
            "Selecione o aviso que deseja desativar:", view=view, ephemeral=True
        )

    # =========================
    # HOURLY LOOP
    # =========================
    @tasks.loop(minutes=1)
    async def hourly_loop(self):
        now = datetime.datetime.now()

        if now.minute != 0:
            return

        if not (12 <= now.hour <= 15):
            return

        index = now.hour - 12
        anns = db.get_active_daily_announcements()

        if index >= len(anns):
            return

        await self.send_single_announcement(anns[index])

    async def send_single_announcement(self, ann):
        channel = self.bot.get_channel(FORCE_CHANNEL_ID)
        if not channel:
            return

        perms = channel.permissions_for(channel.guild.me)
        if not (perms.send_messages and perms.embed_links and perms.attach_files):
            return

        _, text_pt, text_en, img_pt, img_en = ann

        files = []

        embed_pt = discord.Embed(title="📢 Aviso", description=text_pt, color=0x2ECC71)
        embed_en = discord.Embed(title="📢 Notice", description=text_en, color=0x3498DB)

        if is_remote_url(img_pt) and is_remote_url(img_en):
            embed_pt.set_image(url=img_pt)
            embed_en.set_image(url=img_en)
        else:
            files = [
                discord.File(img_pt, filename="pt.png"),
                discord.File(img_en, filename="en.png"),
            ]
            embed_pt.set_image(url="attachment://pt.png")
            embed_en.set_image(url="attachment://en.png")

        await channel.send(embeds=[embed_pt, embed_en], files=files)


async def setup(bot):
    await bot.add_cog(DailyAnnouncement(bot))
